From e48af79372f28c605ed46077733d00b552e8d893 Mon Sep 17 00:00:00 2001
From: Philippe Gerum <philippe.gerum@exail.com>
Date: Fri, 20 Oct 2023 11:37:16 +0200
Subject: [PATCH 854/985] evl/net: fix endianness in protocol matching

The ->attach() handler of a protocol descriptor receives the protocol
number in the wrong (native) endianness. This illustrates a general
issue with the current code: we should not be dealing with protocol
identifiers in network byte order internally, this should be
restricted to external interface code.

Switch all protocol identifiers to native endianness except when
dealing with external interfaces.

Signed-off-by: Philippe Gerum <philippe.gerum@exail.com>
---
 include/evl/net/socket.h       |  6 +++---
 kernel/evl/net/packet/packet.c | 37 ++++++++++++++++------------------
 kernel/evl/net/socket.c        |  7 +++----
 3 files changed, 23 insertions(+), 27 deletions(-)

diff --git a/include/evl/net/socket.h b/include/evl/net/socket.h
index 8a62114c0..651c27c64 100644
--- a/include/evl/net/socket.h
+++ b/include/evl/net/socket.h
@@ -26,7 +26,7 @@ struct net_device;
 
 struct evl_net_proto {
 	int (*attach)(struct evl_socket *esk,
-		struct evl_net_proto *proto, __be16 protocol);
+		struct evl_net_proto *proto, int protocol);
 	void (*detach)(struct evl_socket *esk);
 	int (*bind)(struct evl_socket *esk,
 		struct sockaddr *addr, int len);
@@ -47,7 +47,7 @@ struct evl_net_proto {
 
 struct evl_socket_domain {
 	int af_domain;
-	struct evl_net_proto *(*match)(int type, __be16 protocol);
+	struct evl_net_proto *(*match)(int type, int protocol);
 	struct list_head next;
 };
 
@@ -68,7 +68,7 @@ struct evl_socket {
 	int wmem_max;
 	struct evl_wait_queue wmem_wait;
 	struct evl_crossing wmem_drain;
-	__be16 protocol;
+	int protocol;
 	struct {
 		int real_ifindex;
 		int vlan_ifindex;
diff --git a/kernel/evl/net/packet/packet.c b/kernel/evl/net/packet/packet.c
index 68cf81031..cd765ea29 100644
--- a/kernel/evl/net/packet/packet.c
+++ b/kernel/evl/net/packet/packet.c
@@ -28,8 +28,7 @@
 #include <evl/uaccess.h>
 
 static struct evl_net_proto *
-find_packet_proto(__be16 protocol,
-		struct evl_net_proto *default_proto);
+find_packet_proto(int protocol, struct evl_net_proto *default_proto);
 
 /*
  * Lock nesting: protocol_lock -> rxq->lock -> esk->input_wait.wchan.lock
@@ -48,7 +47,7 @@ static DEFINE_EVL_SPINLOCK(protocol_lock);
 
 /* oob, hard irqs off */
 static bool __packet_deliver(struct evl_net_rxqueue *rxq,
-			struct sk_buff *skb, __be16 protocol)
+			struct sk_buff *skb, int protocol)
 {
 	struct net_device *dev = skb->dev;
 	bool delivered = false;
@@ -100,7 +99,7 @@ static bool __packet_deliver(struct evl_net_rxqueue *rxq,
 		 * consumes the incoming buffer.
 		 */
 		qskb = skb;
-		if (protocol == htons(ETH_P_ALL)) {
+		if (protocol == ETH_P_ALL) {
 			qskb = evl_net_clone_skb(skb);
 			if (qskb == NULL) {
 				evl_flush_wait(&esk->input_wait, EVL_T_NOMEM);
@@ -119,7 +118,7 @@ static bool __packet_deliver(struct evl_net_rxqueue *rxq,
 
 		evl_signal_poll_events(&esk->poll_head,	POLLIN|POLLRDNORM);
 		delivered = true;
-		if (protocol != htons(ETH_P_ALL))
+		if (protocol != ETH_P_ALL)
 			break;
 	}
 
@@ -140,14 +139,14 @@ static struct evl_net_rxqueue *find_rxqueue(u32 hkey)
 	return NULL;
 }
 
-static inline u32 get_protocol_hash(__be16 protocol)
+static inline u32 get_protocol_hash(int protocol)
 {
 	u32 hsrc = protocol;
 
 	return jhash2(&hsrc, 1, 0);
 }
 
-static bool packet_deliver(struct sk_buff *skb, __be16 protocol) /* oob */
+static bool packet_deliver(struct sk_buff *skb, int protocol) /* oob */
 {
 	struct evl_net_rxqueue *rxq;
 	unsigned long flags;
@@ -187,14 +186,14 @@ static bool packet_deliver(struct sk_buff *skb, __be16 protocol) /* oob */
  */
 bool evl_net_packet_deliver(struct sk_buff *skb) /* oob */
 {
-	packet_deliver(skb, htons(ETH_P_ALL));
+	packet_deliver(skb, ETH_P_ALL);
 
-	return packet_deliver(skb, skb->protocol);
+	return packet_deliver(skb, ntohs(skb->protocol));
 }
 
 /* in-band. */
 static int attach_packet_socket(struct evl_socket *esk,
-				struct evl_net_proto *proto, __be16 protocol)
+				struct evl_net_proto *proto, int protocol)
 {
 	struct evl_net_rxqueue *rxq, *_rxq;
 	unsigned long flags;
@@ -285,7 +284,7 @@ static int bind_packet_socket(struct evl_socket *esk,
 	if (sll->sll_family != AF_PACKET)
 		return -EINVAL;
 
-	proto = find_packet_proto(sll->sll_protocol, esk->proto);
+	proto = find_packet_proto(ntohs(sll->sll_protocol), esk->proto);
 	if (proto == NULL)
 		return -EINVAL;
 
@@ -316,7 +315,7 @@ static int bind_packet_socket(struct evl_socket *esk,
 	 */
 	sll->sll_ifindex = real_ifindex;
 
-	if (esk->protocol != sll->sll_protocol) {
+	if (esk->protocol != ntohs(sll->sll_protocol)) {
 		detach_packet_socket(esk);
 		/*
 		 * Since the old binding was dropped, we would not
@@ -325,7 +324,7 @@ static int bind_packet_socket(struct evl_socket *esk,
 		 * root issue would be way more problematic than a
 		 * dead socket.
 		 */
-		ret = attach_packet_socket(esk, proto, sll->sll_protocol);
+		ret = attach_packet_socket(esk, proto, ntohs(sll->sll_protocol));
 		if (ret)
 			goto out;
 	}
@@ -462,7 +461,7 @@ static ssize_t send_packet(struct evl_socket *esk,
 	}
 
 	skb_reset_mac_header(skb);
-	skb->protocol = esk->protocol;
+	skb->protocol = htons(esk->protocol);
 	skb->dev = real_dev;
 	skb->priority = esk->sk->sk_priority;
 
@@ -669,12 +668,11 @@ static struct evl_net_proto ether_packet_proto = {
 };
 
 static struct evl_net_proto *
-find_packet_proto(__be16 protocol,
-		struct evl_net_proto *default_proto)
+find_packet_proto(int protocol,	struct evl_net_proto *default_proto)
 {
 	switch (protocol) {
-	case htons(ETH_P_ALL):
-	case htons(ETH_P_IP):
+	case ETH_P_ALL:
+	case ETH_P_IP:
 		return &ether_packet_proto;
 	case 0:
 		return default_proto;
@@ -683,8 +681,7 @@ find_packet_proto(__be16 protocol,
 	}
 }
 
-static struct evl_net_proto *
-match_packet_domain(int type, __be16 protocol)
+static struct evl_net_proto *match_packet_domain(int type, int protocol)
 {
 	static struct evl_net_proto *proto;
 
diff --git a/kernel/evl/net/socket.c b/kernel/evl/net/socket.c
index e79e9c736..a9c6eb82c 100644
--- a/kernel/evl/net/socket.c
+++ b/kernel/evl/net/socket.c
@@ -354,8 +354,7 @@ void evl_uncharge_socket_wmem(struct sk_buff *skb)
 }
 
 /* in-band */
-static struct evl_net_proto *
-find_oob_proto(int domain, int type, __be16 protocol)
+static struct evl_net_proto *find_oob_proto(int domain, int type, int protocol)
 {
 	struct evl_net_proto *proto = NULL;
 	struct domain_list_head *head;
@@ -455,7 +454,7 @@ int sock_oob_attach(struct socket *sock)
 	esk->wmem_max = sk->sk_sndbuf;
 	evl_init_crossing(&esk->wmem_drain);
 
-	ret = proto->attach(esk, proto, sk->sk_protocol);
+	ret = proto->attach(esk, proto, ntohs(sk->sk_protocol));
 	if (ret)
 		goto fail_attach;
 
@@ -765,7 +764,7 @@ static void destroy_evl_socket(struct sock *sk)
 }
 
 static int create_evl_socket(struct net *net, struct socket *sock,
-			int protocol, int kern)
+			     int protocol, int kern)
 {
 	struct sock *sk;
 
-- 
2.30.2

